File: chakracore-cxx/ffi/Logger.cpp
#include "chakra/Logger.h"
#include "chakracore-sys/src/logger.rs.h"

namespace chakra
{
    void Logger::error(const std::string &message, const std::source_location &location)
    {
        chakra_rs::log::error(location.function_name(), location.file_name(), location.line(), message);
    }

    void Logger::warn(const std::string &message, const std::source_location &location)
    {
        chakra_rs::log::warn(location.function_name(), location.file_name(), location.line(), message);
    }

    void Logger::info(const std::string &message, const std::source_location &location)
    {
        chakra_rs::log::info(location.function_name(), location.file_name(), location.line(), message);
    }

    void Logger::debug(const std::string &message, const std::source_location &location)
    {
        chakra_rs::log::debug(location.function_name(), location.file_name(), location.line(), message);
    }

    void Logger::trace(const std::string &message, const std::source_location &location)
    {
        chakra_rs::log::trace(location.function_name(), location.file_name(), location.line(), message);
    }

    Span::Span(rust::Box<chakra_rs::log::ForeignSpan> foreign_span) : foreign_span_(std::move(foreign_span)) {}

    Span Span::create(const std::string &name)
    {
        auto span = chakra_rs::log::ForeignSpan::create(name);
        return Span{std::move(span)};
    }
} // namespace chakra

File: chakracore-sys/src/logger.rs
use tracing::span::EnteredSpan;

#[cxx::bridge(namespace = "chakra_rs::log")]
mod ffi {
    extern "Rust" {
        fn error(function_name: &str, file_name: &str, line: u32, message: &str);
        fn warn(function_name: &str, file_name: &str, line: u32, message: &str);
        fn info(function_name: &str, file_name: &str, line: u32, message: &str);
        fn debug(function_name: &str, file_name: &str, line: u32, message: &str);
        fn trace(function_name: &str, file_name: &str, line: u32, message: &str);

        type ForeignSpan;
        #[Self = "ForeignSpan"]
        fn create(function_name: &str) -> Box<ForeignSpan>;
    }
}

fn error(function_name: &str, file_name: &str, line: u32, message: &str) {
    tracing::error!(function_name, file_name, line, message);
}

fn warn(function_name: &str, file_name: &str, line: u32, message: &str) {
    tracing::warn!(function_name, file_name, line, message);
}

fn info(function_name: &str, file_name: &str, line: u32, message: &str) {
    tracing::info!(function_name, file_name, line, message);
}

fn debug(function_name: &str, file_name: &str, line: u32, message: &str) {
    tracing::debug!(function_name, file_name, line, message);
}

fn trace(function_name: &str, file_name: &str, line: u32, message: &str) {
    tracing::trace!(function_name, file_name, line, message);
}

struct ForeignSpan {
    _span: EnteredSpan,
}

impl ForeignSpan {
    fn create(function_name: &str) -> Box<ForeignSpan> {
        let span = tracing::info_span!("c++", name = function_name).entered();
        Box::new(ForeignSpan { _span: span })
    }
}
